components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -310,11 +311,84 @@ std::vector<mojom::FeedItemV2Ptr> GenerateChannelBlock(
   }
 
   std::vector<mojom::FeedItemV2Ptr> result;
-  result.push_back(mojom::FeedItemV2::NewCluster(
-      mojom::Cluster::New("channel", channel, std::move(article_elements))));
+  result.push_back(mojom::FeedItemV2::NewCluster(mojom::Cluster::New(
+      mojom::ClusterType::CHANNEL, channel, std::move(article_elements))));
   return result;
 }
 
+std::vector<mojom::FeedItemV2Ptr> GenerateTopicBlock(
+    std::vector<Topic>& topics,
+    const Publishers& publishers) {
+  if (topics.empty()) {
+    return {};
+  }
+  DVLOG(1) << __FUNCTION__;
+  auto result = mojom::Cluster::New();
+  result->type = mojom::ClusterType::TOPIC;
+
+  auto topic = std::move(topics[0]);
+  topics.erase(topics.begin());
+  result->id = topic.title;
+
+  uint64_t max_articles = features::kBraveNewsMaxBlockCards.Get();
+  for (const auto& article : topic.articles) {
+    auto item = mojom::FeedItemMetadata::New();
+    auto id_it = base::ranges::find_if(publishers, [&article](const auto& p) {
+      return p.second->publisher_name == article.publisher_name;
+    });
+    if (id_it != publishers.end()) {
+      item->publisher_id = id_it->first;
+    }
+    item->publisher_name = article.publisher_name;
+    item->category_name = article.category;
+    item->description = article.description;
+    item->title = article.title;
+    item->url = article.url;
+    item->publish_time = base::Time::Now();
+    item->image = mojom::Image::NewImageUrl(article.img);
+    result->articles.push_back(mojom::ArticleElements::NewArticle(
+        mojom::Article::New(std::move(item), false)));
+
+    // For now, we truncate at |max_articles|. In future we may want to include
+    // more articles here and have the option to show more in the front end.
+    if (result->articles.size() >= max_articles) {
+      break;
+    }
+  }
+
+  std::vector<mojom::FeedItemV2Ptr> items;
+  items.push_back(mojom::FeedItemV2::NewCluster(std::move(result)));
+  return items;
+}
+
+std::vector<mojom::FeedItemV2Ptr> GenerateClusterBlock(
```

Comment:
Would it be helpful maybe to have a comment here explaining what this function does, and why it does what it does? It is not entirely clear to me, an unfamiliar eye, why we have the checks down under, with random numbers, and only in certain cases we mutate the contents of `topics`.

Reply:
Add a comment with some basic explanation and a link to the spec document
https://docs.google.com/document/d/1bSVHunwmcHwyQTpa3ab4KRbGbgNQ3ym_GHvONnrBypg/edit#heading=h.agyx2d7gifd9